### PR TITLE
[`flake8-pyi`] Implement `PYI064`

### PR DESCRIPTION
**Pull request by tusharsadhwani**
## Summary

Implements `Y064` from `flake8-pyi` and its autofix.

## Test Plan

`cargo test` / `cargo insta review`

### CODE REVIEW

File: crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+y: Final[Literal[None]] = None  # PYI064
+z: Final[Literal[  # PYI064
+    "this is a really long literal, that won't be rendered in the issue text"
+]] = "this is a really long literal, that won't be rendered in the issue text"
+
+# This should be fixable, and marked as safe
+w1: Final[Literal[123]]  # PYI064
+
+# This should be fixable, but marked as unsafe
+w2: Final[Literal[123]] = "random value"  # PYI064
```
Comment (tusharsadhwani) [inline crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py:13]: I think an unsafe autofix is appropriate here (specifically when the assignment value is different from the literal value).

Is autofixing it to `w2: Final = 123` in unsafe mode fair?

Comment (AlexWaygood) [reply crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py:13]: ummmmmmm this kind of thing doesn't really make any sense, right? I think I'd lean towards not doing an autofix here -- "refuse the temptation to guess"

Comment (tusharsadhwani) [reply crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py:13]: fair enough.

Comment (tusharsadhwani) [reply crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py:13]: done.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_final_literal.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    else {
+        return;
+    };
+    let ast::Expr::Subscript(ast::ExprSubscript { slice: literal, .. }) = &**literal_slice else {
+        return;
+    };
+
+    // Discards tuples like `Literal[1, 2, 3]`
+    // and complex literals like `Literal[{1, 2}]`
+    if !matches!(
```
Comment (tusharsadhwani) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_final_literal.rs:78]: can potentially be moved to `ruff_python_ast/src/helpers.rs`, as a `is_literal` helper.

File: crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# This should not be fixable
+w2: Final[Literal[123]] = "random value"  # PYI064
+
+n1: Final[Literal[True, False]] = True # No issue here
+n2: Literal[True] = True  # No issue here
+
+PlatformName = Literal["linux", "macos", "windows"]
+PLATFORMS: Final[set[PlatformName]] = {"linux", "macos", "windows"}  # No issue here
+
+foo: Final[{1, 2, 3}] = {1, 2, 3}  # No issue here
```
Comment (tusharsadhwani) [inline crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI064.py:21]: Right now this mimics `flake8-pyi` behaviour, however I think raising the issue here should be fine?

**Comment by charliermarsh** [comment]
@AlexWaygood -- will follow-up and merge this.

**Review by charliermarsh — APPROVED** [review]

**Review by Fffjjy — COMMENTED** [review]
Fix

**Pull request merged**